poc_iot_verifier/src/purger.rs

```diff
@@ -0,0 +1,193 @@
+            }
+            tokio::select! {
+                _ = shutdown.clone() => break,
+                _ = db_timer.tick() =>
+                    match self.handle_db_tick(lora_invalid_beacon_tx.clone(),lora_invalid_witness_tx.clone()).await {
+                    Ok(()) => (),
+                    Err(err) => {
+                        tracing::error!("fatal purger error: {err:?}");
+                        return Err(err)
+                    }
+                }
+            }
+        }
+        tracing::info!("stopping purger");
+        Ok(())
+    }
+
+    async fn handle_db_tick(
+        &self,
+        lora_invalid_beacon_tx: MessageSender,
+        lora_invalid_witness_tx: MessageSender,
+    ) -> Result {
+        // pull stale beacons and witnesses which are in a pending state
+        // for each we have to write out an invalid report to S3
+        // as these wont have previously resulted in a file going to s3
+        // once the report is safely on s3 we can then proceed to purge from the db
+        _ = Report::get_stale_pending_beacons(&self.pool)
+            .await?
+            .iter()
+            .map(|report| self.handle_purged_beacon(report, &lora_invalid_beacon_tx));
+
+        _ = Report::get_stale_pending_witnesses(&self.pool)
+            .await?
+            .iter()
+            .map(|report| self.handle_purged_witness(report, &lora_invalid_witness_tx));
+
+        // purge any stale entropy, no need to output anything to s3 here
+        _ = Entropy::purge(&self.pool, &self.entropy_stale_period).await;
+        Ok(())
+    }
+
+    async fn handle_purged_beacon(
+        &self,
+        db_beacon: &Report,
+        lora_invalid_beacon_tx: &MessageSender,
+    ) -> Result {
+        let packet_data = &db_beacon.packet_data;
+        let beacon_buf: &[u8] = &db_beacon.report_data;
+        let beacon_report = LoraBeaconIngestReport::try_from(
+            LoraBeaconIngestReportV1::decode(beacon_buf).unwrap(),
+        )?;
+        let beacon = &beacon_report.report;
+        let invalid_beacon_proto: LoraInvalidBeaconReportV1 = LoraInvalidBeaconReport {
+            received_timestamp: beacon_report.received_timestamp,
+            reason: InvalidReason::Stale,
+            report: beacon.clone(),
+        }
+        .into();
+        file_sink::write(lora_invalid_beacon_tx, invalid_beacon_proto).await?;
+        // delete the report from the DB
+        let public_key = beacon_report.report.pub_key.to_vec();
+        self.delete_db_report(public_key, packet_data.clone()).await;
+        Ok(())
+    }
+
+    async fn handle_purged_witness(
+        &self,
+        db_witness: &Report,
+        lora_invalid_witness_tx: &MessageSender,
+    ) -> Result {
+        let packet_data = &db_witness.packet_data;
+        let witness_buf: &[u8] = &db_witness.report_data;
+        let witness_report = LoraWitnessIngestReport::try_from(
+            LoraWitnessIngestReportV1::decode(witness_buf).unwrap(),
```

Comment:
Match on these and return early if the decode fails instead of unwrapping? Or did we decide these were invariant circumstances?

Reply:
Alrighty; if it’s being addressed 👍